updated RTDCWriter class and tests for image feature

RaghavaAlajangi opened the pull request:
As we discussed before, I changed a few lines of code in RTDCWriter class so that it saves an array of data as an image if the dimensions of the array are greater than three. Moreover, I made corresponding changes in tests.

paulmueller commented:
Thanks for this PR! Please add a test where you define a non-scalar plugin feature and use `RTDCWrite.store_feature` with

1. a 3D array with the shape given in the plugin feature
2. a 2D array with the shape given in the plugin feature
3. 3D array with the shape **not** given in the plugin feature
4. 2D array with the shape **not** given in the plugin feature

RaghavaAlajangi commented:
Would it work as a test for this issue (#171)
```
def test_pf_store_3D_plugin_data_without_shape():
    """Test storage of image plugin feature without shape info in recipe"""
    h5path = retrieve_data("fmt-hdf5_fl_2018.zip")
    ds = dclab.new_dataset(retrieve_data("fmt-hdf5_fl_2018.zip"))
    info = example_plugin_info_non_scalar_feature()
    # info["feature shapes"] = [(80, 250)]
    PlugInFeature("image_gauss_filter", info)

    writer = RTDCWriter(h5path)
    image_gauss_filter = gaussian_filter(ds["image"], sigma=(0, 1, 1))
    writer.store_feature("image_gauss_filter", image_gauss_filter)
    
    with h5py.File(h5path, mode="r") as rtdc_data:
        events = rtdc_data["events"]
        assert "image_gauss_filter" in events.keys()
        image = rtdc_data["events"]['image_gauss_filter']
        assert b'IMAGE' in image.attrs.get('CLASS')
```

paulmueller commented:
Yes, you can use that as a base for the four tests. A few comments:

- you are calling `retrieve_data("fmt-hdf5_fl_2018.zip")` twice. I think this is not necessary.
- please don't put commented code in a test
- you should use `RTDCWriter` in a context (`with RTDCWriter(path) as hw:`)
- you can do the same thing with RTDC dataset (`with dclab,new_dataset(path) as ds: image = ---`)
- instead of `rtdc_data` use `h5` to be consistent with the remainder of the tests

RaghavaAlajangi commented:
> Yes, you can use that as a base for the four tests. A few comments:
> 
> * you are calling `retrieve_data("fmt-hdf5_fl_2018.zip")` twice. I think this is not necessary.
> * please don't put commented-out code in a test
> * you should use `RTDCWriter` in a context (`with RTDCWriter(path) as hw:`)
> * you can do the same thing with RTDC dataset (`with dclab,new_dataset(path) as ds: image = ---`)
> * instead of `rtdc_data` use `h5` to be consistent with the remainder of the tests

Thanks for the comments.

PinkShnack commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -217,7 +217,15 @@ def store_feature(self, feat, data, shape=None):
+                
+            # above code ensure that the shape of the (plugin or temp) data 
+            # must be 3 dimensions --> (None, dim1, dim2).
+            # If dim1 and dim2 are greater than 3, it would be considered as 
```
If **the size of** dim1 and dim2 ...

You could also say (z, y, x) instead of dim2, dim2 as they are a bit comfusing.

RaghavaAlajangi replied:
Instead of (x,y,z), I used rows and columns notation of ndarrays. Is it clear now?

PinkShnack commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -217,7 +217,15 @@ def store_feature(self, feat, data, shape=None):
+            # must be 3 dimensions --> (None, dim1, dim2).
+            # If dim1 and dim2 are greater than 3, it would be considered as 
+            # an image. Otherwise, a ndarray. 
+            if data.shape[1] > 3 and data.shape[2] > 3: 
```
I assume this if statement will cause an index error if the data has less than 2 dimensions? 
It would be good to write a test for this situation.

RaghavaAlajangi replied:
Yes, I just added one more line on top of it to check the length of the data.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -903,7 +904,78 @@ def test_pf_wrong_length_2():
+    """Test storage of 2D image plugin feature without shape info in recipe"""
+    h5path = retrieve_data("fmt-hdf5_fl_2018.zip")
+    info = example_plugin_info_non_scalar_feature()
+    PlugInFeature("dummy_image", info)
```
Maybe have a comment here saying `# we do not add feature shape to the info`

RaghavaAlajangi replied:
Yes, as per your suggestion, I commented over there.

paulmueller commented:
The tests seem to be failing. Could you please take a look?

https://github.com/ZELLMECHANIK-DRESDEN/dclab/runs/5751234710?check_suite_focus=true

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -903,7 +904,78 @@ def test_pf_wrong_length_2():
+        events = h5["events"]
+        assert "dummy_image" in events.keys()
+        image = h5["events"]['dummy_image']
+        assert b'IMAGE' in image.attrs.get('CLASS')
```
In general, it would be good if you created or used an actual 2D plugin feature and tested its use with your new code. See the test "`test_pf_load_plugin()`" for example.

RaghavaAlajangi replied:
That was the cause of the failure of tests. So, I used the existing future function for testing. Now, it's working fine.

PinkShnack commented on tests/test_rtdc_writer.py:
```diff
@@ -66,8 +66,10 @@ def test_bulk_image():
-
-
+        image = rtdc_data["events"]['image']
+        assert b'IMAGE' in image.attrs.get('CLASS')
```
You should also be checking the shape of the image to make sure it is correct.

RaghavaAlajangi replied:
Yes, added shape too

RaghavaAlajangi replied:
> Yes, added shape too

Sorry, I thought this is in another test function.

RaghavaAlajangi replied:
Now, I have updated all these tests in order to check shape of the feature. Please find them in this commit https://github.com/ZELLMECHANIK-DRESDEN/dclab/pull/172/commits/08c8d2097d24f8c1eef7242ed8d27cd283b09300

PinkShnack reviewed (changes requested):
I also see that there are some plugin tests that fail on the ci/cd run. You will have to test these locally and figure out what from your new code broke the old tests!

PinkShnack commented:
> The tests seem to be failing. Could you please take a look?
> 
> https://github.com/ZELLMECHANIK-DRESDEN/dclab/runs/5751234710?check_suite_focus=true

See my review (I think we both posted at the same time 😄 )

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -903,7 +904,88 @@ def test_pf_wrong_length_2():
+        events = h5["events"]
+        assert "image_gauss_filter" in events.keys()
+        feat = h5["events"]['image_gauss_filter']
+        assert b'IMAGE' in feat.attrs.get('CLASS')
```
for all these tests, we want to assert that the shape is as expected in the reloaded hdf5 file, so please check for this.

RaghavaAlajangi replied:
Yes, I have added but haven't committed the changes.

PinkShnack replied:
Did you do this??

RaghavaAlajangi replied:
Yes, you can find them like
```
fshape = feat.shape
assert feat.ndim == 3
```

PinkShnack replied:
I don't see it here, please link the line of code

RaghavaAlajangi replied:
Didn't you mean line number 1002? (check the shape of the feature)

PinkShnack replied:
I have no idea what line of code it is FYI, I just want you to post a link to the solution.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -903,7 +904,88 @@ def test_pf_wrong_length_2():
+        assert "circ_per_area" in events.keys()
+        feat = h5["events"]['circ_per_area']
+        assert b'CLASS' not in feat.attrs.keys()
+        assert b'IMAGE_SUBCLASS' not in feat.attrs.keys()
```
You can definietly add more tests. Paul mentioned 4 possible tests in his https://github.com/ZELLMECHANIK-DRESDEN/dclab/pull/172#issuecomment-1080494432. You can find more edge cases I think!

What happens when you give 4D? are the correct errors raised? What happens if I save an image and define the shape of a contour. Is that possible? If it isn't, show me where :)

RaghavaAlajangi replied:
Yes, you are right. I have to add more tests for every possibility. Right now, I am doing it locally. While doing so, I found that I had a problem with contour data because of its shape `[len, nan, 2]`. I guess we have to make conditions in the `RTDCWriter.store_features()` method to handle every situation (like contour, 2darray, image, 4darray, mask, scalar) of plugin feature. Right now, I just added conditions for image (3D) data only.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -55,6 +55,12 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
 
+def compute_mask_plugin_feature(rtdc_ds):
+    """Basic plugin method creates mask data (boolean data)"""
+    mask_feat = np.logical_not(rtdc_ds["mask"])
```
Maybe use a different name than `mask_feat`. It is too broad.

RaghavaAlajangi replied:
Can you recommend a name?

RaghavaAlajangi replied:
Sorry for the overlook!

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -109,6 +115,24 @@ def example_plugin_info_non_scalar_feature():
+        "long description": "This mask feature is the mask "
+                            "data of RTDC",
+        "feature names": ["mask_feat"],
+        "feature labels": ["Mask or boolean Image"],
```
This is intended as the "unit" of the feature. For example, for `area_um`, its feature label is "um" (micrometers). In this case there are not units, like here: https://dclab.readthedocs.io/en/stable/sec_av_notation.html#non-scalar-features

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -924,14 +948,18 @@ def test_pf_store_non_scalar_plugin_data_with_shape_issue_171():
-def test_pf_store_non_scalar_plugin_data_without_shape_issue_171():
-    """Test storage of 3D image plugin feature attributes without shape info
-     in recipe"""
+def test_pf_store_image_plugin_data_without_shape_issue_171():
```
Don't link the name of the function to the issue. The function name + short docstring description should be enough to understand what it is doing.

RaghavaAlajangi replied:
It is fixed now.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -924,14 +948,18 @@ def test_pf_store_non_scalar_plugin_data_with_shape_issue_171():
-     in recipe"""
+def test_pf_store_image_plugin_data_without_shape_issue_171():
+    """Test storage of image plugin feature attributes without shape info
+     in the recipe"""
```
Remove "in the recipe"

PinkShnack replied:
Same for the other docstrings

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -943,7 +971,59 @@ def test_pf_store_non_scalar_plugin_data_without_shape_issue_171():
         assert "image_gauss_filter" in events.keys()
         feat = h5["events"]['image_gauss_filter']
+        fshape = feat.shape
+        assert len(fshape) == 3 and fshape[1] > 3 and fshape[2] > 3
```
You can have these assertions on separate lines for better clarity. Also, what shape is dim[0]?

PinkShnack replied:
Same comment for the other tests

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -964,6 +1044,8 @@ def test_pf_store_scalar_plugin_data_with_shape_issue_171():
         assert "circ_per_area" in events.keys()
         feat = h5["events"]['circ_per_area']
+        fshape = feat.shape
+        assert len(fshape) == 1
```
I think a nicer way to look at the number of dimensions might be `assert feat.ndim == 1`.

PinkShnack reviewed (commented):
Some small comments. Tests look great!

PinkShnack commented:
The only thing stopping the CI/CD pipeline is some linting problems (flake8).

PinkShnack commented:
@RaghavaAlajangi you will need to fix the ci/cd pipeline before we can review your code! Or if you have questions ask away.

RaghavaAlajangi commented:
@PinkShnack please review my code and let me know your comments.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+@pytest.mark.filterwarnings(
+    "ignore::dclab.rtdc_dataset.config.WrongConfigurationTypeWarning")
+def test_pf_store_mask_plugin_data_without_shape_issue_171():
+    """Test storage of mask plugin feature attributes without shape info"""
```
without shape info? You are setting it in line 1044

RaghavaAlajangi replied:
I dont get your point here. I am just passing an argument `shape` of the feature in `store_feature()`.

PinkShnack replied:
In [line 1044](https://github.com/RaghavaAlajangi/dclab/blob/master/tests/test_rtdc_feat_anc_plugin.py#L1044) you are passing shape info, so the description is not right imo

RaghavaAlajangi replied:
Ha..I get it. Thanks

RaghavaAlajangi replied:
I changed the function name to address this (added `shape_in_info` at the end of function)

RaghavaAlajangi replied:
I changed the function name accordingly (added `shape_in_info()`). Is it clear enough?

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+
+@pytest.mark.filterwarnings(
+    "ignore::dclab.rtdc_dataset.config.WrongConfigurationTypeWarning")
+def test_pf_store_mask_plugin_data_without_shape_issue_171():
```
issue 171 still in name

RaghavaAlajangi replied:
It is fixed now.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+    """Test storage of ragged plugin feature"""
+    h5path = retrieve_data("fmt-hdf5_fl_2018.zip")
+    info = example_plugin_info_ragged_data_feature()
+    with dclab.new_dataset(h5path) as ds:
```
You are repeating the code from the plugin function here, why?

RaghavaAlajangi replied:
It is fixed now.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+    with dclab.new_dataset(h5path) as ds:
+        cnts = ds["contour"]  # shape --> [None, nan, 2]
+        # Add dummy contour coordinates to the actual contours
+        ragged_feat = np.array([np.vstack((c, np.ones((3, 2)))) for c in cnts],
```
This you should be getting from the plugin feature.

RaghavaAlajangi replied:
But in `store_feature()` , we have to pass `data` argument right? where can I get the data from?

RaghavaAlajangi replied:
I fixed this, please have a look!

RaghavaAlajangi replied:
It is fixed now.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+
+@pytest.mark.filterwarnings(
+    "ignore::dclab.rtdc_dataset.config.WrongConfigurationTypeWarning")
+def test_pf_store_scalar_plugin_data_issue_171():
```
Best to put the scalar functions above the non-scalar, makes more sense in my brain at least

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+    """Test storage of scalar plugin feature attributes"""
+    h5path = retrieve_data("fmt-hdf5_fl_2018.zip")
+    with dclab.new_dataset(h5path) as ds:
+        circ_per_area = ds["circ"] / ds["area_um"]
```
Why are you calling the data like this when you have the plugin feature?

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+    info = example_plugin_info_non_scalar_feature()
+    PlugInFeature("image_gauss_filter", info)
+    with dclab.new_dataset(h5path) as ds:
+        image_gauss_filter = gaussian_filter(ds["image"], sigma=(0, 1, 1))
```
same problem here

RaghavaAlajangi replied:
It is fixed now.

PinkShnack commented on tests/test_rtdc_writer.py:
```diff
@@ -89,6 +91,8 @@ def test_bulk_mask():
         # Masks are stored as uint8
         assert np.allclose(events["mask"][6], mask[6]*255)
         assert events["mask"][1].shape == (20, 10)
```
Check the shape of `events["mask"]` also.

PinkShnack commented on tests/test_rtdc_writer.py:
```diff
@@ -350,11 +354,9 @@ def test_non_scalar_bad_shape():
-        # This should work
-        hw.store_feature("peterpan", data, shape=(3, 5))
-        # This should not work
+        data = data.reshape(1, *data.shape)
```
Maybe describe what you are doing here in a comment?

RaghavaAlajangi replied:
Actually, this step is already covered in `write_image_grayscale()` method (see line number 370). So, I will remove this line from here. This line basically add an extra dimension served as an index in HDFview.

RaghavaAlajangi replied:
It is fixed now. I removed the line 357 because it is not needed.

PinkShnack reviewed (changes requested):
Thanks for all the changes. Here are some changes that are needed :D

RaghavaAlajangi commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -209,15 +208,32 @@ def store_feature(self, feat, data, shape=None):
                         + f"you could put the shape `{data[0].shape}` "
                         + 'in the `info["feature shapes"]` key of '
                         + "your plugin feature.")
-                    shape = data.shape[1:]
```
Actually, no. I have to keep this line. I have added it.

RaghavaAlajangi replied:
It is fixed now.

RaghavaAlajangi commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -209,15 +208,32 @@ def store_feature(self, feat, data, shape=None):
-            elif shape == data.shape[1:]:
-                pass
+            # Condition for ragged/contour data (array of arrays)
+            if data.ndim == 1 and isinstance(data[0], np.ndarray):
```
No, contour data is actually a list of arrays so it has `data.ndim = 1`.

PinkShnack replied:
But is `data` in this case a list? In that case it won't have an ndim property, no?

RaghavaAlajangi replied:
I didnt see lists in dclab to store the data. I guess we are only using arrays because they are much faster than lists.

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -54,6 +55,40 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
+    # Ragged data (i.e. list of arrays of different lengths) --> Ex.contours
+    cnts = rtdc_ds["contour"]  # shape --> [None, nan, 2]
+    # Add dummy contour coordinates to the actual contours
+    ragged_feat = np.array([np.vstack((c, np.ones((3, 2)))) for c in cnts],
```
I guess we need to derive pf from the existing features right? with keep that in my mind, I just added some dummy coordinates to the contour data.

RaghavaAlajangi replied:
Yes, you are right! whereas, I just want to put some dummy data just to differentiate between normal and pf data. If you dont recommend that I will remove it.

PinkShnack replied:
It is fine the way it is! :)

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -54,6 +55,40 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
+def compute_ragged_plugin_feature(rtdc_ds):
+    """Basic plugin method creates arrays of different lengths (ragged)"""
+    # Ragged data (i.e. list of arrays of different lengths) --> Ex.contours
+    cnts = rtdc_ds["contour"]  # shape --> [None, nan, 2]
```
We dont need that comment. I can remove that!

PinkShnack replied:
You should verify the shape anyway I think

RaghavaAlajangi replied:
You mean assertion?

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -163,9 +216,9 @@ def test_pf_attributes():
+           plugin_file_info["feature names"][0]
     assert pf2.feature_name == pf2.feature_name == \
-        plugin_file_info["feature names"][1]
+           plugin_file_info["feature names"][1]
```
Ohh...I didnt realize it. Thank you :)

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
 
+@pytest.mark.filterwarnings(
+    "ignore::dclab.rtdc_dataset.config.WrongConfigurationTypeWarning")
+def test_pf_store_image_plugin_data_with_shape_issue_171():
```
I thought it's common that we need to mention the issue number. I removed the issue number from all of my test functions.

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +958,160 @@ def test_pf_wrong_length_2():
+        assert fshape[1] > 3 and fshape[2] > 3
+        assert b'CLASS' in feat.attrs.keys()
+        assert b'IMAGE_SUBCLASS' in feat.attrs.keys()
+        assert b'IMAGE' in feat.attrs.get('CLASS')
```
Yes, I checked with HDFView.

RaghavaAlajangi replied:
You can find this here (https://github.com/ZELLMECHANIK-DRESDEN/dclab/issues/171#issue-1178295751)

RaghavaAlajangi replied:
It is fixed now.

RaghavaAlajangi commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -209,15 +208,32 @@ def store_feature(self, feat, data, shape=None):
+
+            # Condition for array of scalar features
+            elif data.ndim == 1 and not isinstance(data[0], np.ndarray):
+                self.write_ndarray(group=events, name=feat, data=data)
```
This is happening because scalar PF's are not saved by this line. They are saved by the condition above (line number 174 to 177). I got to know this via debugging. Thank you for telling me how to make use of debugging!

PinkShnack replied:
The issue is that the line isn't being tested though, so when you have fixed this, please resolve this thread

RaghavaAlajangi replied:
It is fixed now.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +964,186 @@ def test_pf_wrong_length_2():
+    info["feature shapes"] = [(80, 250)]
+    PlugInFeature("mask_other", info)
+    with dclab.new_dataset(h5path) as ds:
+        mask_other = np.logical_not(ds["mask"])
```
You are not using plugin features correctly FYI. Please check out the plugin feature explanation on dclab.

RaghavaAlajangi replied:
I fixed it. You can see the changes here (https://github.com/ZELLMECHANIK-DRESDEN/dclab/pull/172/commits/326e58ac6dc1df5391586c71c8c8c1b5db9f934e)

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -54,6 +55,46 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
+        "method": compute_ragged_plugin_feature,
+        "description": "This plugin will derive ragged feature from existing "
+                       "contour data by simply adding some dummy coordinates",
+        "long description": "Ragged data or irregular data is nothing but a "
```
Yes, I will removed.

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +964,186 @@ def test_pf_wrong_length_2():
+        assert not isinstance(feat[0], np.ndarray)
+        assert not feat.attrs.keys()
+
+
```
Have you done this?

RaghavaAlajangi replied:
You can find the screenshot here (https://github.com/ZELLMECHANIK-DRESDEN/dclab/pull/172#issuecomment-1081930071)

PinkShnack replied:
Aha, cool. Maybe actually place it in the original issue so that general users can see it! And a screenshot of the actual image open in HDFView too!

RaghavaAlajangi replied:
Yes, you can find here (https://github.com/ZELLMECHANIK-DRESDEN/dclab/issues/171#issue-1178295751)

PinkShnack commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +964,186 @@ def test_pf_wrong_length_2():
+    with h5py.File(h5path, mode="r") as h5:
+        events = h5["events"]
+        assert "ragged_feat" in events.keys()
+        assert np.allclose(events["ragged_feat"]["1"], ragged_feat[1])
```
Apparently this is normal.

RaghavaAlajangi replied:
it is fixed now.

PinkShnack commented:
@RaghavaAlajangi, did you comment "I fixed this, please have a look!" and then delete the comment? Or is it somewhere else? Please link to it.

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -54,6 +55,46 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
+    cnts = rtdc_ds["contour"]
+    # Add dummy contour coordinates to the actual contours
+    # dummy = [np.vstack((c, np.ones((3, 2)))) for c in cnts]
+    dummy = [c for c in cnts]
```
Actually, the contour data is `LazyContourList` object. We cannot use this object directly so I tried to create a ragged array by adding additional rows. 

Yes, I need to uncomment the line 69. I forgot to do that.

RaghavaAlajangi replied:
It is fixed now.

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -54,6 +55,46 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
+                            "list of arrays of different shapes. For example, "
+                            "in contour feature, each contour of the cell has "
+                            "different number of coordinates. So we cannot "
+                            "save this feature as a regular matrix instead of "
```
Yes

RaghavaAlajangi commented:
> @RaghavaAlajangi, did you comment "I fixed this, please have a look!" and then delete the comment? Or is it somewhere else? Please link to it.

Yes, I did it but I thought it would be good to comment after committing changes. So, I deleted it. 

Now, I committed all the changes. I guess I fixed all the issues that you mentioned before. Please let me know if you find any new issues.

RaghavaAlajangi commented on tests/test_rtdc_writer.py:
```diff
@@ -35,6 +35,7 @@ def test_bulk_contour():
     for ii in range(5, num + 5):
         cii = np.arange(2 * ii).reshape(2, ii)
         contour.append(cii)
+    print(len(contour))
```
I forgot to delete this line.

RaghavaAlajangi replied:
Is fixed now.

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -54,6 +55,40 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
+    """plugin info for non-image feature"""
+    info = {
+        "method": compute_ragged_plugin_feature,
+        "description": "This plugin will compute a ragged feature",
```
It is fixed now.

paulmueller commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -210,14 +216,33 @@ def store_feature(self, feat, data, shape=None):
+                # Condition for single image or array of shape (H, W). If the
+                # given data is a 2D, it is more likely a single image feature
+                # therefore it will be saved with image attributes.
+                elif data.ndim == 2:
```
Shouldn't you be checking against `shape` here and not `data.ndim`? The warning message above would otherwise be invalid. What if someone just stores a new scalar (temporary) feature? Then `data.ndim` is two but shape is `(1,)`.

RaghavaAlajangi replied:
True, will change this.

paulmueller commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -210,14 +216,33 @@ def store_feature(self, feat, data, shape=None):
-            elif shape == data.shape[1:]:
-                pass
+            # Condition for single or multiple events to be saved
+            if shape == data.shape or shape == data.shape[1:]:
```
It would be much cleaner if you kept the original separation between single and multiple events. Combining them in one case complicates things a lot.

paulmueller replied:
How about introducing an `is_ragged` variable? Then, after these two original cases you decide based on the `shape` whether to write as ragged, image_grayscale, or ndarray (because you may also have 3d features)?

RaghavaAlajangi replied:
> It would be much cleaner if you kept the original separation between single and multiple events. Combining them in one case complicates things a lot.

In that case, we would end up with a lot of conditions, right?

paulmueller commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -210,14 +216,33 @@ def store_feature(self, feat, data, shape=None):
+                if data.ndim == 1 and isinstance(data[0], np.ndarray):
+                    self.write_ragged(group=events, name=feat, data=data)
+                # Condition for single image or array of shape (H, W). If the
+                # given data is a 2D, it is more likely a single image feature
```
This comment is misleading. "more likely" than a ragged array?

RaghavaAlajangi replied:
Sorry for the inconsistency with the commenting. It has to be under the below case (2D image).

paulmueller commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -210,14 +216,33 @@ def store_feature(self, feat, data, shape=None):
+                                               data=data,
+                                               is_boolean=(dtype != bool))
+                # Condition for list of images or array of shape (None, H, W).
+                elif data.ndim == 3:
```
Here, again, you should be checking against `shape`. I believe this is too complicated.

paulmueller commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +962,183 @@ def test_pf_wrong_length_2():
+        feat = h5["events"]['mask_other']
+        assert feat.ndim == 3
+        fshape = feat.shape
+        assert fshape[1] > 3 and fshape[2] > 3
```
Why not check for the exact shape (80, 250)? (See also other tests)

paulmueller commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +962,183 @@ def test_pf_wrong_length_2():
+        im_shape = image_gauss_filter.shape
+        ext_image_gauss_filter = image_gauss_filter.reshape(1, *im_shape)
+    with RTDCWriter(h5path) as hw:
+        with pytest.raises(ValueError, match="Bad shape"):
```
Why do you expect a ValueError here? The feature could be a 3D feature, correct? I would expect the original warning message about assuming a certain shape.

RaghavaAlajangi replied:
Actually, the feature shape would be 4D here (added extra dimension line 1136) because I thought of making a limitation to the function that only saves data of shape below 4D.

I am not sure whether it is a good addition or not (please see the condition in [here]( https://github.com/RaghavaAlajangi/dclab/blob/08c8d2097d24f8c1eef7242ed8d27cd283b09300/dclab/rtdc_dataset/writer.py#L240))

paulmueller reviewed (changes requested):
Thanks for all the work! This thing really has blown up. You need to simplify this (especially combining the two cases in the writer made the code much less readable).

As far as I can see, you have not implemented all the test cases I proposed (which should all pass):

1. Write a 3D array with the shape given in the plugin feature
2. Write a 2D array with the shape given in the plugin feature
3. Write 3D array with the shape not given in the plugin feature (should only warn)
4. Write a 2D array with the shape not given in the plugin feature (should only warn)

Just to be sure that no functionality is broken and there is not limit regarding dimensions, I would now also like to see a test case for 4D as well (representative for ND).

For the ragged data, a plugin feature should be able to specify a shape with `nan`. I wonder whether that is supported right now. If not, then maybe it makes sense to split this PR up into smaller parts and do that first. It might also make sense to write the feature shape as an attribute to the HDF5 dataset (or group, in case of ragged features) and let the HDF5 reader handle that properly.

RaghavaAlajangi commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -165,6 +164,13 @@ def store_feature(self, feat, data, shape=None):
             self.write_ndarray(group=events,
                                name="index",
                                data=np.arange(nev0 + 1, nev0 + nev + 1))
+        # This condition checks whether the given feature is a scalar or not.
```
Yes, I will change this.

RaghavaAlajangi commented on dclab/rtdc_dataset/writer.py:
```diff
@@ -165,6 +164,13 @@ def store_feature(self, feat, data, shape=None):
                                data=np.arange(nev0 + 1, nev0 + nev + 1))
+        # This condition checks whether the given feature is a scalar or not.
+        # If it is a scalar feature, it will be saved in the rtdc file
+        # It also takes the scalar plugin feature into account because during
```
Aha...ok. I will replace it with `instantiation`.

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -54,6 +55,44 @@ def compute_non_scalar_plugin_feature(rtdc_ds):
+
+def compute_ragged_plugin_feature(rtdc_ds):
+    """Basic plugin method creates arrays of different lengths (ragged)"""
+    # Ragged data (i.e. list of arrays of different shapes) --> Ex.contours
```
Example!!

RaghavaAlajangi replied:
If it is not clear enough, I will change it.

RaghavaAlajangi commented on tests/test_rtdc_feat_anc_plugin.py:
```diff
@@ -905,6 +962,183 @@ def test_pf_wrong_length_2():
 
+@pytest.mark.filterwarnings(
+    "ignore::dclab.rtdc_dataset.config.WrongConfigurationTypeWarning")
+def test_pf_store_scalar_plugin_data_with_shape_in_info():
```
You are right! basically, this test function is for saving a scalar feature with/without the shape given in the Plugin feature.

RaghavaAlajangi commented on dclab/definitions/feat_logic.py:
```diff
@@ -18,7 +19,9 @@ def check_feature_shape(name, data):
         If the data's shape does not match its scalar description
     """
-    if len(data.shape) == 1 and not scalar_feature_exists(name):
+    # For ragged data (contour like), additional condition has been added.
```
You are right! but I wanted to tell you what had been changed over there. Do you think that the comment is necessary?

paulmueller commented:
I am closing this PR since the code base has changed quite significantly since then. Feel free to create a new PR on this isue.
